bookkeeper-server/src/main/java/org/apache/bookkeeper/client/BookKeeperBuilderImpl.java

```diff
@@ -0,0 +1,94 @@
+ * under the License.
+ *
+ */
+package org.apache.bookkeeper.client;
+
+import com.google.common.base.Preconditions;
+import io.netty.channel.EventLoopGroup;
+import io.netty.util.HashedWheelTimer;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import static org.apache.bookkeeper.client.BKException.Code.ZKException;
+import org.apache.bookkeeper.client.api.BookKeeper;
+import org.apache.bookkeeper.client.api.BookKeeperBuilder;
+import org.apache.bookkeeper.conf.ClientConfiguration;
+import org.apache.bookkeeper.feature.FeatureProvider;
+import org.apache.bookkeeper.net.DNSToSwitchMapping;
+import org.apache.bookkeeper.stats.NullStatsLogger;
+import org.apache.bookkeeper.stats.StatsLogger;
+import org.apache.zookeeper.KeeperException;
+import org.apache.zookeeper.ZooKeeper;
+
+/**
+ * Internal builder for {@link org.apache.bookkeeper.client.api.BookKeeper} client
+ */
+public class BookKeeperBuilderImpl implements BookKeeperBuilder {
+
+    private final ClientConfiguration conf;
+    private ZooKeeper zk = null;
+    private EventLoopGroup eventLoopGroup;
+    private StatsLogger statsLogger = NullStatsLogger.INSTANCE;
+    private DNSToSwitchMapping dnsResolver;
+    private HashedWheelTimer requestTimer;
+    private FeatureProvider featureProvider;
+
+    public BookKeeperBuilderImpl(ClientConfiguration conf) {
+        this.conf = conf;
+    }
+
+    @Override
+    public BookKeeperBuilder with(Object component) {
+        Preconditions.checkNotNull(component);
+        if (component instanceof ZooKeeper) {
+            this.zk = (ZooKeeper) component;
+        } else if (component instanceof EventLoopGroup) {
+            this.eventLoopGroup = (EventLoopGroup) component;
+        } else if (component instanceof StatsLogger) {
+            this.statsLogger = (StatsLogger) component;
+        } else if (component instanceof DNSToSwitchMapping) {
+            this.dnsResolver = (DNSToSwitchMapping) component;
+        } else if (component instanceof HashedWheelTimer) {
+            this.requestTimer = (HashedWheelTimer) component;
+        } else if (component instanceof FeatureProvider) {
+            this.featureProvider = (FeatureProvider) component;
+        } else {
+            throw new IllegalArgumentException("Component of type " + component.getClass().getName()
+                    + " is not configurable");
+        }
+        return this;
+    }
+
+    @Override
+    public BookKeeper build() throws BKException, InterruptedException {
+        try {
+            return new org.apache.bookkeeper.client.BookKeeper(conf, zk, eventLoopGroup,
+                    statsLogger, dnsResolver, requestTimer, featureProvider);
+        }
+        catch (KeeperException err) {
+            BKException.ZKException zkErr = new BKException.ZKException();
+            zkErr.initCause(err);
+            throw zkErr;
+        }
+        catch (IOException err) {
+            throw new UncheckedIOException(err);
```

Comment:
why this is UncheckedIOException?

Reply:
because I wanted to hide the IOException.
There is no good way to a have a BKException for "client creation".
I am going to explicitly add a "throws" clause to the builder